File: Python/4kyu/Human Readable Duration Format/solution.py
# Yeah, I admit.
# I took the lazy approach with this one.
# It was incredibly frustrating trying to make it work in
# python. Well, I've since then learned my lesson and
# created a more "automatic" solution in the other programming
# languages instead of brute forcing my solution.

def format_duration(seconds):
    if seconds == 0:
        result = "now"

    years = seconds//31536000
    days = (seconds-(31536000*years))//86400
    hours = (seconds - years*31536000 - days*86400)//3600
    minutes = (seconds - years*31536000 - days*86400 - hours*3600)//60
    seconds = seconds - years*31536000 - days*86400 - hours*3600 - minutes*60
    
    if years and days and hours and minutes and seconds:
        result = ("{0} year{1}, ".format(years, "s" if years!=1 else "") if years else "") + \
        ("{0} day{1}, ".format(days, "s" if days!=1 else "") if days else "") + \
        ("{0} hour{1}, ".format(hours, "s" if hours!=1 else "") if hours else "") + \
        ("{0} minute{1} and ".format(minutes, "s" if minutes!=1 else "") if minutes else "") + \
        ("{0} second{1}".format(seconds, "s" if seconds!=1 else "") if seconds else "")
    elif years and days and hours and minutes:
        result = ("{0} year{1}, ".format(years, "s" if years!=1 else "") if years else "") + \
        ("{0} day{1}, ".format(days, "s" if days!=1 else "") if days else "") + \
        ("{0} hour{1} and ".format(hours, "s" if hours!=1 else "") if hours else "") + \
        ("{0} minute{1}".format(minutes, "s" if minutes!=1 else "") if minutes else "")
    elif days and hours and minutes and seconds:
        result = ("{0} day{1}, ".format(days, "s" if days!=1 else "") if days else "") + \
        ("{0} hour{1}, ".format(hours, "s" if hours!=1 else "") if hours else "") + \
        ("{0} minute{1} and ".format(minutes, "s" if minutes!=1 else "") if minutes else "") + \
        ("{0} second{1}".format(seconds, "s" if seconds!=1 else "") if seconds else "")
    elif hours and minutes and seconds:
        result = ("{0} hour{1}, ".format(hours, "s" if hours!=1 else "") if hours else "") + \
        ("{0} minute{1} and ".format(minutes, "s" if minutes!=1 else "") if minutes else "") + \
        ("{0} second{1}".format(seconds, "s" if seconds!=1 else "") if seconds else "")
    elif days and minutes and seconds:
        result = ("{0} day{1}, ".format(days, "s" if days!=1 else "") if days else "") + \
        ("{0} minute{1} and ".format(minutes, "s" if minutes!=1 else "") if minutes else "") + \
        ("{0} second{1}".format(seconds, "s" if seconds!=1 else "") if seconds else "")
    elif days and hours and minutes:
        result = ("{0} day{1}, ".format(days, "s" if days!=1 else "") if days else "") + \
        ("{0} hour{1} and ".format(hours, "s" if hours!=1 else "") if hours else "") + \
        ("{0} minute{1}".format(minutes, "s" if minutes!=1 else "") if minutes else "")
    elif days and hours and seconds:
        result = ("{0} day{1}, ".format(days, "s" if days!=1 else "") if days else "") + \
        ("{0} hour{1} and ".format(hours, "s" if hours!=1 else "") if hours else "") + \
        ("{0} second{1}".format(seconds, "s" if seconds!=1 else "") if seconds else "")    
    elif minutes and seconds:
        result = ("{0} minute{1} and ".format(minutes, "s" if minutes!=1 else "") if minutes else "") + \
        ("{0} second{1}".format(seconds, "s" if seconds!=1 else "") if seconds else "")
    elif seconds:
        result = ("{0} second{1}".format(seconds, "s" if seconds!=1 else "") if seconds else "")
    elif minutes:
        result = ("{0} minute{1}".format(minutes, "s" if minutes!=1 else "") if minutes else "")
    elif hours:
        result = ("{0} hour{1}".format(hours, "s" if hours!=1 else "") if hours else "")
    elif days:
        result = ("{0} day{1}".format(days, "s" if days!=1 else "") if days else "")

    return result

File: C/4kyu/Human Readable Duration Format/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char *formatDuration (int n) {
  const char *times[] = {"year", "day", "hour", "minute", "second"};
  unsigned long int timesValues[] = {365 * 24 * 60 * 60, 24 * 60 *60, 60 * 60, 60, 1};
  int timesResults[5];
  int size = 0;
  char *string = calloc(57, sizeof(char));
  char buffer[40];
  char date[8];
  
  if(n == 0)
    return strcpy(string, "now");

  timesResults[0] = n/timesValues[0];
  timesResults[1] = (n-timesValues[0]*timesResults[0])/timesValues[1];
  timesResults[2] = (n-timesValues[0]*timesResults[0]-timesValues[1]*timesResults[1])/timesValues[2];
  timesResults[3] = (n-timesValues[0]*timesResults[0]-timesValues[1]*timesResults[1]-timesValues[2]*timesResults[2])/timesValues[3];
  timesResults[4] = n-timesValues[0]*timesResults[0]-timesValues[1]*timesResults[1]-timesValues[2]*timesResults[2]-timesValues[3]*timesResults[3];

  // Get the size of "valid" times
  for(int i = 0; i < 5; i++) {
    if(timesResults[i] > 0)
      size++;
  }

  // Creating the string to be returned
  for(int i = 0; i < 5; i++) {
    if(timesResults[i] > 0) {
      --size;
      strcpy(date, times[i]);

      if(timesResults[i] > 1)
        strcat(date, "s");

      sprintf(buffer, "%d %s", timesResults[i], date);
      strcat(string, buffer);

      if(size > 1)
        strcat(string, ", ");
      else if (size == 1)
        strcat(string, " and ");
    }
  }

  return string;
}